### PR TITLE
Fix message nonce tracking during head changes

### PR DESCRIPTION
**Pull request by whyrusleeping**
Also wrote up a test framework for the message pool, should allow us to write nicer tests. Only wrote the one to repro the issue for now.

### CODE REVIEW

File: chain/messagepool/messagepool.go [hunk 1]
```diff
@@ -307,16 +355,44 @@ func (mp *MessagePool) getNonceLocked(addr address.Address) (uint64, error) {
...
+	act, err := mp.api.StateGetActor(addr, mp.curTs)
 	if err != nil {
 		return 0, err
 	}
 
-	return act.Nonce, nil
+	baseNonce := act.Nonce
+
+	// TODO: the correct thing to do here is probably to set curTs to chain.head
+	// but since we have an accurate view of the world until a head change occurs,
```
Comment (magik6k) [inline chain/messagepool/messagepool.go]: I'm not convinced that this is actually true

File: chain/messagepool/messagepool.go [hunk 2]
```diff
@@ -98,20 +105,61 @@ func (ms *msgSet) add(m *types.SignedMessage) error {
...
+
+func (mpp *mpoolProvider) MessagesForBlock(h *types.BlockHeader) ([]*types.Message, []*types.SignedMessage, error) {
+	return mpp.sm.ChainStore().MessagesForBlock(h)
+}
+
+func (mpp *mpoolProvider) MessagesForTipset(ts *types.TipSet) ([]store.ChainMsg, error) {
+	return mpp.sm.ChainStore().MessagesForTipset(ts)
+}
+
+func NewMessagePool(api MpoolProvider, ds dtypes.MetadataDS) (*MessagePool, error) {
```
Comment (magik6k) [inline chain/messagepool/messagepool.go]:
```suggestion
func New(api MpoolProvider, ds dtypes.MetadataDS) (*MessagePool, error) {
```

File: chain/messagepool/messagepool.go [hunk 3]
```diff
@@ -98,20 +105,61 @@ func (ms *msgSet) add(m *types.SignedMessage) error {
...
+	MessagesForBlock(*types.BlockHeader) ([]*types.Message, []*types.SignedMessage, error)
+	MessagesForTipset(*types.TipSet) ([]store.ChainMsg, error)
+}
+
+type mpoolProvider struct {
+	sm *stmgr.StateManager
+	ps *pubsub.PubSub
+}
+
+func NewMpoolProvider(sm *stmgr.StateManager, ps *pubsub.PubSub) MpoolProvider {
```
Comment (magik6k) [inline chain/messagepool/messagepool.go]:
```suggestion
func NewProvider(sm *stmgr.StateManager, ps *pubsub.PubSub) MpoolProvider {
```

File: chain/messagepool/messagepool.go [hunk 4]
```diff
@@ -98,20 +105,61 @@ func (ms *msgSet) add(m *types.SignedMessage) error {
 	return nil
 }
 
-func NewMessagePool(sm *stmgr.StateManager, ps *pubsub.PubSub, ds dtypes.MetadataDS) (*MessagePool, error) {
+type MpoolProvider interface {
```
Comment (magik6k) [inline chain/messagepool/messagepool.go]:
```suggestion
type Provider interface {
```

**Comment by whyrusleeping** [comment]
tests hanging for me for some reason...

**Pull request merged**